rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,417 @@
+
+
+class RainbowCommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = router_address
+
+    def _enricher_event(
+            self,
+            event: 'EvmEvent',
+            sender: ChecksumEvmAddress,
+    ) -> None:
+        """Set the event to rainbow event
+        """
+        if (
+            event.location_label != sender or
+            event.address != self.router_address or
+            event.counterparty is not None
+        ):
+            return
+        elif event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE:  # noqa: E501
+            # Find the receive event
+            event.event_type = HistoryEventType.TRADE
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.counterparty = CPT_RAINBOW_SWAPS
+            event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+        elif event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE:  # noqa: E501
+            # Find the spend event
+            event.event_type = HistoryEventType.TRADE
+            event.event_subtype = HistoryEventSubType.SPEND
+            event.notes = f'Swap {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap'  # noqa: E501
+            event.counterparty = CPT_RAINBOW_SWAPS
+            event.address = self.router_address
+
+    def _find_event_by_log(
+            self,
+            logs: list['EvmTxReceiptLog'],
+            event_signature: bytes,
+    ) -> EvmTxReceiptLog | None:
+        """
+        Find the log by the event signature.
+
+        Args:
+            logs (list[EvmTxReceiptLog]): List of logs.
+            event_signature (bytes): Event signature to be searched.
+
+        Returns:
+            Optional[EvmTxReceiptLog]: Log if found, else None.
+        """
+        for tx_log in logs:
+            if tx_log.topics[0] == event_signature:
+                return tx_log
+        return None
+
+    def _get_in_out_events(
+            self,
+            events: list['EvmEvent'],
+            sender: ChecksumEvmAddress,
+    ) -> tuple[Optional['EvmEvent'], Optional['EvmEvent']]:
+        """
+        Gets the in and out event in the list of events
+        Args:
+            events (list['EvmEvent']): A list of EVM events to be processed.
+            sender (ChecksumEvmAddress): The address of the sender involved in the events.
+        Returns:
+            in_event: Represents the receive the tokens from the swap
+            ou_event: Represents the spend the tokens to swap
```

Comment:
Typo in out_events but also we don't follow this docstring style. I agree that in this one it gives good context but keep it simple

```python
"""
        Extracts 'in' and 'out' events for a sender from a list of events.

        Args:
            events: List of EVM events.
            sender: Sender address to filter events by.

        Returns:
            Tuple of (in_event, out_event), representing token receive and spend events respectively.
        """
```